File: inject-java/src/test/java/io/micronaut/aop/around/noproxytarget/ArgMutatingInterceptor.java
package io.micronaut.aop.around.noproxytarget;

import io.micronaut.aop.Interceptor;
import io.micronaut.aop.InvocationContext;
import io.micronaut.core.type.MutableArgumentValue;
import jakarta.inject.Singleton;

@Singleton
public class ArgMutatingInterceptor implements Interceptor {

    @Override
    public Object intercept(InvocationContext context) {
        Mutating m = context.synthesize(Mutating.class);
        MutableArgumentValue arg = (MutableArgumentValue) context.getParameters().get(m.value());
        if (arg != null) {
            Object value = arg.getValue();
            if (value instanceof Number number) {
                arg.setValue(number.intValue() * 2);
            } else {
                arg.setValue("changed");
            }
        }
        return context.proceed();
    }
}

File: inject-kotlin/src/test/kotlin/io/micronaut/kotlin/processing/aop/proxytarget/ArgMutatingInterceptor.kt
package io.micronaut.kotlin.processing.aop.proxytarget

import io.micronaut.aop.Interceptor
import io.micronaut.aop.InvocationContext
import io.micronaut.core.type.MutableArgumentValue
import jakarta.inject.Singleton

@Singleton
class ArgMutatingInterceptor : Interceptor<Any, Any> {

    override fun intercept(context: InvocationContext<Any, Any>): Any? {
        val m = context.stringValue(
            Mutating::class.java
        )
        if (m.isPresent) {
            val arg = context.parameters[m.get()] as MutableArgumentValue<Any>?
            if (arg != null) {
                val value = arg.value
                if (value is Number) {
                    arg.setValue((value.toInt() * 2))
                } else {
                    arg.setValue("changed")
                }
            }
        }
        return context.proceed()
    }
}
